rust/lance-encoding/src/encodings/physical/bitpack.rs

```diff
@@ -0,0 +1,650 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::sync::Arc;
+
+use arrow::array::ArrayData;
+use arrow::datatypes::{ArrowPrimitiveType, UInt16Type, UInt32Type, UInt64Type, UInt8Type};
+use arrow::util::bit_util::ceil;
+use arrow_array::{cast::AsArray, Array, ArrayRef, PrimitiveArray};
+use arrow_schema::DataType;
+use bytes::{Bytes, BytesMut};
+use futures::future::{BoxFuture, FutureExt};
+use log::trace;
+use num_traits::{AsPrimitive, PrimInt};
+use snafu::{location, Location};
+
+use lance_arrow::DataTypeExt;
+use lance_core::{Error, Result};
+
+use crate::encoder::EncodedBufferMeta;
+use crate::{
+    decoder::{PageScheduler, PrimitivePageDecoder},
+    encoder::{BufferEncoder, EncodedBuffer},
+};
+
+// Compute the number of bits to use for each item, if this array can be encoded using
+// bitpacking encoding. Returns `None` if the type or array data is not supported.
+pub fn num_compressed_bits(arr: ArrayRef) -> Option<u64> {
+    match arr.data_type() {
+        DataType::UInt8 => Some(num_bits_for_type::<UInt8Type>(arr.as_primitive())),
+        DataType::UInt16 => Some(num_bits_for_type::<UInt16Type>(arr.as_primitive())),
+        DataType::UInt32 => Some(num_bits_for_type::<UInt32Type>(arr.as_primitive())),
+        DataType::UInt64 => Some(num_bits_for_type::<UInt64Type>(arr.as_primitive())),
+        // TODO -- eventually we could support signed types as well
```

Comment:
Can do in a follow-up but we also want the various temporal types to be encoded with bit packing too.